### PR TITLE
Remove getProperties from ShareFileClient.downloadToFile()

### LINKED ISSUE #4527: Remove getProperties call for share file download to file

### PR DESCRIPTION
**Pull request by ibrahimrabab**
# Description

resolves #4527

Previously ShareFileAsyncClient.downloadToFile() made an unnecessary call of .getProperties().To avoid this, we are downloading using a single call to downloadWithResponse(). 

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ModelHelper.java [hunk 1]
```diff
@@ -196,4 +200,28 @@ public static ShareFileItemProperties transformFileProperty(FileProperty propert
...
+            headers.getCacheControl(),
+            headers.getContentDisposition(),
+            headers.getCopyCompletionTime(),
+            headers.getCopyStatusDescription(),
+            headers.getCopyId(),
+            headers.getCopyProgress(),
+            headers.getCopySource(),
+            headers.getCopyStatus(),
+            headers.isServerEncrypted(),
+            null);
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ModelHelper.java:224]: One minor worry is customers who were using `FileSmbProperties` from download before will no longer have access to them.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java [hunk 2]
```diff
@@ -964,38 +964,33 @@ public Mono<Response<ShareFileProperties>> downloadToFileWithResponse(String dow
...
+            .retryWhen(Retry.max(3).filter(throwable -> throwable instanceof IOException
+                || throwable instanceof TimeoutException));
     }
 
+    private Mono<Response<ShareFileProperties>> downloadSingleChunk(AsynchronousFileChannel channel, ShareFileRange range,
+        ShareRequestConditions requestConditions, Context context) {
+        ShareFileRange finalRange = range == null ? new ShareFileRange(0) : range;
+        return downloadWithResponse(new ShareFileDownloadOptions().setRange(finalRange)
+            .setRequestConditions(requestConditions), context)
+            .subscribeOn(Schedulers.boundedElastic())
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java:987]: nit: the `subscribeOn` isn't necessary

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java [hunk 3]
```diff
@@ -964,38 +964,33 @@ public Mono<Response<ShareFileProperties>> downloadToFileWithResponse(String dow
...
+            .writeFile(response.getValue(), channel, chunkRange.getStart() - (range == null ? 0 : range.getStart()))
+            .subscribeOn(Schedulers.boundedElastic())
+            .retryWhen(Retry.max(3).filter(throwable -> throwable instanceof IOException
+                || throwable instanceof TimeoutException));
     }
 
+    private Mono<Response<ShareFileProperties>> downloadSingleChunk(AsynchronousFileChannel channel, ShareFileRange range,
+        ShareRequestConditions requestConditions, Context context) {
+        ShareFileRange finalRange = range == null ? new ShareFileRange(0) : range;
+        return downloadWithResponse(new ShareFileDownloadOptions().setRange(finalRange)
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java:985]: If possible, this should get wrapped with `FluxUtil.createRetriableDownloadFlux` so the download can be resumed from a midpoint if it fails.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java [hunk 4]
```diff
@@ -964,38 +964,33 @@ public Mono<Response<ShareFileProperties>> downloadToFileWithResponse(String dow
...
-            .then(Mono.just(response));
+
+    private static Mono<Void> writeBodyToFile(ShareFileDownloadAsyncResponse response, AsynchronousFileChannel channel,
+        ShareFileRange chunkRange, ShareFileRange range) {
+
+        return FluxUtil
+            .writeFile(response.getValue(), channel, chunkRange.getStart() - (range == null ? 0 : range.getStart()))
+            .subscribeOn(Schedulers.boundedElastic())
+            .retryWhen(Retry.max(3).filter(throwable -> throwable instanceof IOException
+                || throwable instanceof TimeoutException));
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java:979]: Similar comment where these aren't needed. We should let better retrying mechanisms handle this.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java [hunk 5]
```diff
@@ -964,38 +964,33 @@ public Mono<Response<ShareFileProperties>> downloadToFileWithResponse(String dow
...
-                .map(ShareFileDownloadAsyncResponse::getValue)
-                .subscribeOn(Schedulers.boundedElastic())
-                .flatMap(fbb -> FluxUtil
-                    .writeFile(fbb, channel, chunk.getStart() - (range == null ? 0 : range.getStart()))
-                    .subscribeOn(Schedulers.boundedElastic())
-                    .retryWhen(Retry.max(3).filter(throwable -> throwable instanceof IOException
-                        || throwable instanceof TimeoutException))))
-            .then(Mono.just(response));
+
+    private static Mono<Void> writeBodyToFile(ShareFileDownloadAsyncResponse response, AsynchronousFileChannel channel,
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java:972]: We may want to expose this through a new API, similar to what Blobs is doing where the Response type is something custom. Callers into the download without Response can call through this as well. That way we can offer a better choice but retain functionality of `ShareFileProperties`, and we won't need to add the translation code.

**Comment by ibrahimrabab** [comment]
Closing this PR since it is not possible to extract FileSmbProperties from the downloadWithResponse call. getProperties() call is needed to ensure we extract all relevant information needed by the user.

**Pull request closed without merging**